From 68d72a88d5dc983aff2ff5dce65581de0ee42b6d Mon Sep 17 00:00:00 2001
From: Michael Schroeder <mls@suse.de>
Date: Wed, 5 Jun 2019 15:58:55 +0200
Subject: [PATCH] Make libsolv work with rpm 5.2

Version 5.2 still used native endian encoding for the database ids.

[Additional Modification in Upstream patch]:
Issue:
libsolv needs to build with -DRPM5=on which is
required to build hawkey.
The rpm-4 doesn't have rpmversion.h file which gives
RPMLIB_VERSION and related macro.
There is Index Out of bound exception is occuring because
database ID is exceeding 2^31

Solution:
Below Upstream Patches fix this issue with rpm-5:
https://github.com/openSUSE/libsolv/commit/68d72a88d5dc983aff2ff5dce65581de0ee42b6d
https://github.com/openSUSE/libsolv/commit/841287dfb7a5992b374aa5326e776cd7d54fc0da
https://github.com/openSUSE/libsolv/commit/ab54abba2ba4eac2fcf0f35cf05e61fb070212eb

So, i modified the upstream patch to work with rpm-4

Signed-off-by: Ankit Jain <ankitja@vmware.com>
- 
---
 ext/repo_rpmdb_bdb.h | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/ext/repo_rpmdb_bdb.h b/ext/repo_rpmdb_bdb.h
index ed82a69..e4329a8 100644
--- a/ext/repo_rpmdb_bdb.h
+++ b/ext/repo_rpmdb_bdb.h
@@ -22,7 +22,11 @@
 #endif
 
 #ifdef RPM5
-# define RPM_INDEX_SIZE 4	/* just the rpmdbid */
+# ifdef RPM5_BIG_ENDIAN_ID
+#  define RPM_INDEX_SIZE 4	/* just the rpmdbid */
+# else
+#  define RPM_INDEX_SIZE 8	/* rpmdbid + array index */
+# endif
 #else
 # define RPM_INDEX_SIZE 8	/* rpmdbid + array index */
 #endif
@@ -70,7 +74,7 @@ stat_database(struct rpmdbstate *state, char *dbname, struct stat *statbuf, int
 static inline Id
 db2rpmdbid(unsigned char *db, int byteswapped)
 {
-#ifdef RPM5
+#ifdef RPM5_BIG_ENDIAN_ID
   return db[0] << 24 | db[1] << 16 | db[2] << 8 | db[3];
 #else
 # if defined(WORDS_BIGENDIAN)
@@ -87,7 +91,7 @@ db2rpmdbid(unsigned char *db, int byteswapped)
 static inline void
 rpmdbid2db(unsigned char *db, Id id, int byteswapped)
 {
-#ifdef RPM5
+#ifdef RPM5_BIG_ENDIAN_ID
   db[0] = id >> 24, db[1] = id >> 16, db[2] = id >> 8, db[3] = id;
 #else
 # if defined(WORDS_BIGENDIAN)
